### PR TITLE
Include checkbox field labels in metadata

### PR DESCRIPTION
**Pull request by ezraporter**
# Description
This PR updates our processing of the REDCap metadata to include checkbox option labels in the `field_label` column for each checkbox option row. The output differs slightly from what I described in #70:

Input:
|field_name|field_label|select_choices_or_calculations|
|----------------------|----------------------|-------------------------|
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: | cyclophosphamide, Cyclophosphamide \| atg, ATG \| other, Other {crs_anti_tcell_therapy_types_specify}|

Output:
|field_name|field_label|updated_field_name|
|----------------------|----------------------|-------------------------|
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: (Cyclophosphamide) | crs_anti_tcell_therapy_types___cyclophosphamide |
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: (ATG) | crs_anti_tcell_therapy_types___atg |
crs_anti_tcell_therapy_types | Anti-T Cell therapy given: (Other) | crs_anti_tcell_therapy_types___other |

Two questions:
1. Not sure I like the carried over ":" in the middle of the field label but not sure how much additional formatting we should do by default. Should we do more? Currently we're just stripping field embedding + whitespace from the checkbox labels.
2. I had to make a decision on what happens if the checkbox field itself doesn't have a label. The current behavior is to print an explicit `"NA"` in the place of the field label before the checkbox option label. So:

|field_name|field_label|select_choices_or_calculations|
|---------|---------|----------|
my_field |  `NA` | 1, Option Label |

becomes

|field_name|field_label|updated_field_name|
|-------|-----------|--------------|
my_field |  NA (Option Label) |  my_field___1 |

Are there alternative suggestions?

# Proposed Changes

- Alter `update_field_names()` to edit `db_metadata$field_label` in place by appending checkbox option labels after the field label
    - This involved moving the `parse_labels()` call from `checkbox_appender()` into the body of `update_field_names()` so we can use both the labels and raw values. At that point enough of `checkbox_appender()` was whittled away that I just removed it.

### Issue Addressed
Addresses #70 (not closing until merged to main with #65)

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -169,77 +169,94 @@ parse_labels <- function(string, return_vector = FALSE) {
...
+      clean_names <- paste(
+        out$field_name[[i]],
+        tolower(parsed_labs$raw),
+        sep = "___"
+      )
+
+      # Strip embedding logic and white space before appending to field labels
+      clean_labs <- parsed_labs$label %>%
+        # TODO: replace with formatting functions?
+        str_replace_all("\\{.*?\\}", "") %>%
```
Comment (rsh52) [inline R/utils.R:229]: Did we confirm that this should be a default enforced option? Is there any reason we may want to have embedding/{} available, or added to the default `formatter` options? In my mind if we're not defaulting stripping HTML tags then we shouldn't default stripping embedded fields either.

**Review by rsh52 — COMMENTED** [review]
Just one comment, otherwise I think this looks good and behaves well. Will definitely want the option to remove HTML once the `formatter` is ready. Great stuff!

File: R/utils.R (hunk 1 above)
Comment (ezraporter) [reply R/utils.R:229]: I think @skadauke and I hastily made that decision while meeting yesterday but I don't have strong opinions. If we chose not to edit the checkbox option labels on the way into the metadata the user would still be able to do things like strip html/embedding logic on the way into `labelled` labels when they call `make_labelled()`.

Comment (rsh52) [reply R/utils.R:229]: I'm could see going either way since they serve different purposes, but I tend to lean toward showing the user everything and then letting them decide what they want to take away (or giving them the option to do so).

Comment (skadauke) [reply R/utils.R:229]: To clarify - we're not talking about formatting labels for `labelled` functionality here but about changing the label of an option of a radio/checkbox field (although this also ultimately affects the tibble's variable labels). I believe that you _always_ want to remove field embedding code here.

**Comment by skadauke** [comment]
> Two questions:
> 
> 1. Not sure I like the carried over ":" in the middle of the field label but not sure how much additional formatting we should do by default. Should we do more? Currently we're just stripping field embedding + whitespace from the checkbox labels.

Agree re: not carrying over `:`. I think what generally looks good is to have the field name followed by a colon and then the option label. It will look bad if there's extra whitespace and/or punctuation (`.`, `:`, `?`) here. I think, specifically for the purpose of generating checkbox field labels, we might want to strip extra whitespace and punctuation at the end of the field label, add a `:`, and then add the option label.

Not sure I'm a fan of enclosing the option label in parens. I think always having a colon between the field label and option label makes this unnecessary.

> 2. I had to make a decision on what happens if the checkbox field itself doesn't have a label. The current behavior is to print an explicit `"NA"` in the place of the field label before the checkbox option label. 

Great point. I think the best bet here will be to use information from the field name. We could remove underlines and make title case.  

So:

field_name	| field_label |	select_choices_or_calculations
----------|-----------------|------
my_field |	`NA` |	1, Option Label

field_name	 |field_label |	updated_field_name
----------|-----------------|------
my_field	| My Field: Option Label |	my_field___1

Do y'all think `make_labelled` should throw a warning if there are fields that don't have field labels?

File: R/utils.R (hunk 1 above)
Comment (ezraporter) [reply R/utils.R:229]: Correct. The question is what processing we should do to create the contents of `supertbl$redcap_metadata` before `labelled` gets involved

Comment (skadauke) [reply R/utils.R:229]: It's a little bit confusing, but now that I'm thinking about it more there I believe that there are two different scenarios where field embedding code might pop up and they may need to be handled differently.

The first is a multi-checkbox such as in the example above. Here we'd want `Other {crs_anti_tcell_therapy_types_specify}` to be `Other`, and if this was the only scenario this could reasonably be handled with a formatter function.

The second scenario in which field embeddings can come up is in the context of a radio field (and possibly also dropdown but I'm not sure). In this case, `Other {something_specify}` doesn't affect the `labelled` variable field. But it _does_ affect the factor label of the variable which will be `Other {something_specify}` instead of `Other`.

Comment (ezraporter) [reply R/utils.R:229]: Should we just give the user the option of applying a formatting function to all checkbox/radio/dropdown labels as an argument in `read_redcap_tidy()` or is that overly complex?

If yes we could centralize the logic in `parse_labels()` so that it would flow into both the metadata field labels (for checkboxes) and the factor levels (for everything else).

**Comment by ezraporter** [comment]
> Agree re: not carrying over `:`. I think what generally looks good is to have the field name followed by a colon and then the option label. It will look bad if there's extra whitespace and/or punctuation (`.`, `:`, `?`) here. I think, specifically for the purpose of generating checkbox field labels, we might want to strip extra whitespace and punctuation at the end of the field label, add a `:`, and then add the option label.
> 
> Not sure I'm a fan of enclosing the option label in parens. I think always having a colon between the field label and option label makes this unnecessary.

I'm okay with this pending the discussion here re processing of options labels https://github.com/CHOP-CGTDataOps/REDCapTidieR/pull/73#discussion_r1021882354. Thoughts, @rsh52?

**Comment by ezraporter** [comment]
> I think the best bet here will be to use information from the field name. We could remove underlines and make title case.  
> 
> So:
> 
> field_name	| field_label |	select_choices_or_calculations
> ----------|-----------------|------
> my_field |	`NA` |	1, Option Label
> 
> field_name	 |field_label |	updated_field_name
> ----------|-----------------|------
> my_field	| My Field: Option Label |	my_field___1

Using the variable name seems like a good idea. I wonder if doing the processing of removing underscores and converting to title case makes too many assumptions about the user's field names.

> Do y'all think make_labelled should throw a warning if there are fields that don't have field labels?

My gut reaction is that warning for this is too noisy. Mainly because:
1. Many REDCaps I've seen contain all kinds of hidden helper fields that you wouldn't want or need labels for
2. It's visually apparent in the viewer when labels are missing

I guess my combined personal answers to these questions are:
1. Use the variable name unedited for checkbox fields with no field label
    - Doing this will also keep it more visually clear when a label was missing but I'm also open to warning/messaging about this
    - i.e. `my_field is a checkbox field with no field label. Converting field labels to format "my_field: {checkbox option label}"`
2. In the general case of fields with no field labels I don't think we should warn

**Comment by skadauke** [comment]
> > I think the best bet here will be to use information from the field name. We could remove underlines and make title case.
> > So:
> > 
> > 
> > 
> > field_name
> > field_label
> > select_choices_or_calculations
> > 
> > 
> > 
> > 
> > my_field
> > `NA`
> > 1, Option Label
> > 
> > 
> > 
> > 
> > 
> > 
> > field_name
> > field_label
> > updated_field_name
> > 
> > 
> > 
> > 
> > my_field
> > My Field: Option Label
> > my_field___1
> 
> Using the variable name seems like a good idea. I wonder if doing the processing of removing underscores and converting to title case makes too many assumptions about the user's field names.
> 
> > Do y'all think make_labelled should throw a warning if there are fields that don't have field labels?
> 
> My gut reaction is that warning for this is too noisy. Mainly because:
> 
> 1. Many REDCaps I've seen contain all kinds of hidden helper fields that you wouldn't want or need labels for
> 2. It's visually apparent in the viewer when labels are missing
> 
> I guess my combined personal answers to these questions are:
> 
> 1. Use the variable name unedited for checkbox fields with no field label
>    
>    * Doing this will also keep it more visually clear when a label was missing but I'm also open to warning/messaging about this
>    * i.e. `my_field is a checkbox field with no field label. Converting field labels to format "my_field: {checkbox option label}"`
> 2. In the general case of fields with no field labels I don't think we should warn

How about this: If field_label is `NA`, you don't get a variable label at all?

File: R/utils.R (hunk 1 above)
Comment (ezraporter) [reply R/utils.R:229]: I'm okay with this. I kind of agree with @rsh52's suggestion above https://github.com/CHOP-CGTDataOps/REDCapTidieR/pull/73#discussion_r1020283174 (which he may or may not still stand by) that if we're stripping embedding logic we also strip html.

I can't think of a reason a user would want raw html in their factor levels

Comment (rsh52) [reply R/utils.R:229]: Yea, it's still my opinion that applying these changes should be done synonymously even if they serve different purposes. I could see the same person that says "I want to see the HTML" also saying "I want to see the field embedding". It may not be the most useful thing in 99% of cases but it's still a true representation of the output and may have its place somewhere. Some REDCap aficionado may be confused when things they expect to be there no longer are.

**Comment by ezraporter** [comment]
We decided to:
1. Strip html and field embedding logic from all field labels and option labels by default in `read_redcap_tidy()`
2. Leave checkbox labels blank if `field_label` is `NA`
3. Not issue warnings for missing field labels

Will rebase on `main`, implement those changes, and open a new PR.
